File: android_app/app/src/main/java/com/apiaryguard/app/MainActivity.java
package com.apiaryguard.app;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

/**
 * Główna aktywność aplikacji ApiaryGuard do monitorowania uli
 */
public class MainActivity extends AppCompatActivity {
    
    private RecyclerView recyclerView;
    private ApiaryAdapter adapter;
    private SwipeRefreshLayout swipeRefreshLayout;
    private TextView statusTextView;
    private FloatingActionButton refreshButton;
    private MainViewModel viewModel;
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        
        // Inicjalizacja widoków
        initViews();
        
        // Konfiguracja ViewModel
        viewModel = new ViewModelProvider(this).get(MainViewModel.class);
        
        // Konfiguracja RecyclerView
        setupRecyclerView();
        
        // Konfiguracja SwipeRefreshLayout
        setupSwipeRefresh();
        
        // Konfiguracja przycisku odświeżania
        setupRefreshButton();
        
        // Obserwowanie danych z ViewModel
        observeData();
        
        // Ładowanie danych
        loadApiaries();
    }
    
    private void initViews() {
        recyclerView = findViewById(R.id.recyclerView);
        swipeRefreshLayout = findViewById(R.id.swipeRefreshLayout);
        statusTextView = findViewById(R.id.statusTextView);
        refreshButton = findViewById(R.id.refreshButton);
    }
    
    private void setupRecyclerView() {
        adapter = new ApiaryAdapter(this);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adapter);
    }
    
    private void setupSwipeRefresh() {
        swipeRefreshLayout.setOnRefreshListener(() -> {
            loadApiaries();
        });
        swipeRefreshLayout.setColorSchemeResources(
            R.color.primary,
            R.color.primary_dark,
            R.color.accent
        );
    }
    
    private void setupRefreshButton() {
        refreshButton.setOnClickListener(v -> loadApiaries());
    }
    
    private void observeData() {
        viewModel.getApiaries().observe(this, apiaries -> {
            if (apiaries != null && !apiaries.isEmpty()) {
                adapter.setApiaries(apiaries);
                statusTextView.setVisibility(View.GONE);
                recyclerView.setVisibility(View.VISIBLE);
            } else {
                statusTextView.setText("Brak danych o ulach");
                statusTextView.setVisibility(View.VISIBLE);
                recyclerView.setVisibility(View.GONE);
            }
        });
        
        viewModel.getLoadingState().observe(this, isLoading -> {
            swipeRefreshLayout.setRefreshing(isLoading);
        });
        
        viewModel.getError().observe(this, error -> {
            if (error != null) {
                Toast.makeText(this, error, Toast.LENGTH_LONG).show();
                viewModel.clearError();
            }
        });
    }
    
    private void loadApiaries() {
        viewModel.loadApiaries();
    }
}

File: android_app/app/src/main/java/com/apiguard/apiary/ui/MainActivity.kt
package com.apiguard.apiary.ui

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.apiguard.apiary.R
import com.apiguard.apiary.databinding.ActivityMainBinding
import com.apiguard.apiary.model.ApiaryData

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var viewModel: MainViewModel
    private lateinit var adapter: ApiaryAdapter
    
    // Klucze do zapisywania stanu
    companion object {
        private const val KEY_SCROLL_POSITION = "scroll_position"
        private const val KEY_DIALOG_SHOWING = "dialog_showing"
    }
    
    // Zmienna do śledzenia czy dialog jest wyświetlany
    private var isIpDialogShowing = false
    private var scrollPosition = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupViewModel()
        setupRecyclerView()
        setupClickListeners()
        
        // Przywróć stan z savedInstanceState jeśli istnieje
        savedInstanceState?.let {
            scrollPosition = it.getInt(KEY_SCROLL_POSITION, 0)
            isIpDialogShowing = it.getBoolean(KEY_DIALOG_SHOWING, false)
            
            // Przywróć pozycję przewijania
            if (scrollPosition > 0) {
                binding.recyclerView.scrollToPosition(scrollPosition)
            }
            
            // Reset flagi dialogu - Observer i tak wywoła showIpInputDialog() jeśli potrzeba
        }
        
        // Sprawdź czy mamy zapisany adres IP przy każdym uruchomieniu (również po odtworzeniu stanu)
        // Nowy MainViewModel zaczyna z nieustawionym stanem połączenia, więc musimy sprawdzić zapisane dane
        viewModel.checkSavedConnection()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        
        // Zapisz pozycję przewijania listy
        val layoutManager = binding.recyclerView.layoutManager as? LinearLayoutManager
        scrollPosition = layoutManager?.findFirstVisibleItemPosition() ?: 0
        outState.putInt(KEY_SCROLL_POSITION, scrollPosition)
        
        // Zapisz stan dialogu
        outState.putBoolean(KEY_DIALOG_SHOWING, isIpDialogShowing)
    }

    private fun setupViewModel() {
        viewModel = ViewModelProvider(this, 
            MainViewModelFactory(applicationContext)
        )[MainViewModel::class.java]

        viewModel.connectionState.observe(this) { isConnected ->
            if (!isConnected) {
                showIpInputDialog()
            }
        }

        viewModel.isLoading.observe(this) { isLoading ->
            binding.swipeRefresh.isRefreshing = isLoading
        }

        viewModel.errorMessage.observe(this) { error ->
            error?.let {
                Toast.makeText(this, it, Toast.LENGTH_LONG).show()
                viewModel.clearError()
            }
        }

        viewModel.apiaryData.observe(this) { data ->
            adapter.submitList(data)
            binding.emptyState.visibility = if (data.isEmpty()) View.VISIBLE else View.GONE
            binding.recyclerView.visibility = if (data.isEmpty()) View.GONE else View.VISIBLE
        }
    }

    private fun setupRecyclerView() {
        adapter = ApiaryAdapter { apiaryData ->
            showApiaryDetails(apiaryData)
        }
        
        binding.recyclerView.apply {
            layoutManager = LinearLayoutManager(this@MainActivity)
            adapter = this@MainActivity.adapter
        }
    }

    private fun setupClickListeners() {
        binding.swipeRefresh.setOnRefreshListener {
            viewModel.loadData()
        }
        
        binding.btnChangeIp.setOnClickListener {
            showIpInputDialog()
        }
    }

    private fun showIpInputDialog() {
        // Jeśli dialog już jest wyświetlany, nie pokazuj ponownie (zapobiega wyciekowi przy rotacji)
        if (isIpDialogShowing) return
        
        val savedIp = viewModel.getSavedIpAddress() ?: ""
        val savedPort = viewModel.getSavedPort().toString()
        
        val view = layoutInflater.inflate(R.layout.dialog_ip_input, null)
        val ipEditText = view.findViewById<android.widget.EditText>(R.id.edit_ip_address)
        val portEditText = view.findViewById<android.widget.EditText>(R.id.edit_port)
        
        ipEditText.setText(savedIp)
        portEditText.setText(savedPort)
        
        AlertDialog.Builder(this)
            .setTitle("Konfiguracja połączenia")
            .setMessage("Podaj statyczny adres IP Raspberry Pi:")
            .setView(view)
            .setPositiveButton("Połącz") { _, _ ->
                val ipAddress = ipEditText.text.toString().trim()
                val portText = portEditText.text.toString()
                
                // Walidacja adresu IP
                if (ipAddress.isEmpty()) {
                    Toast.makeText(this, "Podaj poprawny adres IP", Toast.LENGTH_SHORT).show()
                    isIpDialogShowing = false
                    return@setPositiveButton
                }
                
                // Walidacja formatu IP
                if (!isValidIpAddress(ipAddress)) {
                    Toast.makeText(this, "Nieprawidłowy format adresu IP", Toast.LENGTH_SHORT).show()
                    isIpDialogShowing = false
                    return@setPositiveButton
                }
                
                // Walidacja portu
                val port = portText.toIntOrNull()
                if (port == null || port < 1 || port > 65535) {
                    Toast.makeText(this, "Port musi być w zakresie 1-65535", Toast.LENGTH_SHORT).show()
                    isIpDialogShowing = false
                    return@setPositiveButton
                }
                
                viewModel.verifyAndSaveConnection(ipAddress, port)
            }
            .setNegativeButton("Anuluj") { _, _ ->
                isIpDialogShowing = false
            }
            .setOnDismissListener {
                isIpDialogShowing = false
            }
            .show()
        
        // Oznacz że dialog jest wyświetlany
        isIpDialogShowing = true
    }

    /**
     * Waliduje format adresu IPv4
     */
    private fun isValidIpAddress(ip: String): Boolean {
        val ipPattern = Regex(
            "^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}" +
            "(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$"
        )
        return ipPattern.matches(ip)
    }

    private fun showApiaryDetails(apiaryData: ApiaryData) {
        AlertDialog.Builder(this)
            .setTitle(apiaryData.name)
            .setMessage(
                "Temperatura: ${apiaryData.temperature}°C\n" +
                "Wilgotność: ${apiaryData.humidity}%\n" +
                "Waga: ${apiaryData.weight}kg\n" +
                "Bateria: ${apiaryData.battery}%"
            )
            .setPositiveButton("OK", null)
            .show()
    }
}
